lib/service/base.rb

```diff
@@ -257,6 +261,39 @@ def run_step
       end
     end
 
+    # @!visibility private
+    class LockStep < Step
+      include StepsHelpers
+
+      attr_reader :steps
+
+      def initialize(*keys, &block)
+        @keys = keys
+        @steps = []
+        instance_exec(&block)
+      end
+
+      def run_step
+        success =
+          DistributedMutex.synchronize(lock_name) do
+            steps.each { |step| step.call(instance, context) }
+            :success
```

Comment:
Here we’re returning `:success` to check if the lock was acquired, right? I see that [`DistributedMutex#get_lock`](https://github.com/discourse/discourse/blob/main/lib/distributed_mutex.rb#78) can raise a `Discourse::ReadOnly` exception too. (I’m not sure acquiring the lock can fail if redis is not in readonly mode, as it seems it’s just waiting for the lock to expire :thinking:)
Anyway, I guess we could handle both things like that:
```ruby
  DistributedMutex.synchronize(lock_name) do
    …
    :success
  end.then { raise LockNotAcquired unless _1 == :success }
rescue LockNotAcquired, Discourse::ReadOnly
  context[result_key].fail(lock_not_acquired: true)
  context.fail!
```
Wdyt?